### ISSUE TITLE
Only a subset of commands available if vim-go is installed in Vim8 packages opt directory

### ISSUE THREAD

**Issue by vendion** [fatih/vim-go#2633]
### What did you do? (required: The issue will be **closed** when not provided)

Install `vim-go` as an optional package using Vim8 Pacages (e.g. ~/.vim/pack/dist/opt/vim-go) rather than in the _start_ directory.

### What did you expect to happen?

I expected `vim-go` to function normally, with all commands available.

### What happened instead?

Only the following commands work: `GoInstallBinaries`, `GoPath`, `GoTests`, `GoTestsAll`, `GoUpdateBinaries`. If I move `vim-go` from the _opt_ directory into the _start_ directory everything works as expected.
 
#### vim-go version:

vim-go: v1.21

####  `vimrc` you used to reproduce (use a *minimal* vimrc with other plugins disabled; do not link to a 2,000 line vimrc):

The relevent part can be found here https://gitlab.com/vendion/dotfiles/tree/opt-vim-go/.vim/pack/dist where `vim-go` exists as a git submodule in the _opt_ directory.

#### Vim version (first three lines from `:version`):
```
VIM - Vi IMproved 8.1 (2018 May 18, compiled Dec 17 2019 11:21:55)
Included patches: 1-2372
Compiled by root@120amd64-default-job-05
```

#### Go version (`go version`):
go version go1.13.5 freebsd/amd64

#### Go environment
<details><summary><code>go env</code> Output:</summary><br><pre>
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/vendion/.cache/go-build"
GOENV="/home/vendion/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="freebsd"
GONOPROXY=""
GONOSUMDB=""
GOOS="freebsd"
GOPATH="/home/vendion/gocode"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/freebsd_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build381136967=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

**Comment by bhcleek** [comment]
There isn't enough information here for me to tell how to duplicate this issue. I'll need to know:
1. When and how did you execute `packload vim-go`?
2. At what point are you trying to see vim-go commands (e.g. before or after opening a Go file)?

**Comment by vendion** [comment]
> When and how did you execute packload vim-go?

I tried executing it 3 different ways, the end goal was if I'm not working on a Go project then vim should skip this plugin. The first way was having `packadd vim-go` in _$HOME/.vim/ftplugins/go.vim_, when that didn't work I put it is _$HOME/.vim/after/ftplugns/go.vim_ (like how it currently is in https://gitlab.com/vendion/dotfiles/blob/opt-vim-go/.vim/after/ftplugin/go.vim), lastly when neither of those worked I tried calling it manually from within vim.

> At what point are you trying to see vim-go commands (e.g. before or after opening a Go file)?

I tried calling them after opening a Go file, by passing the file name as an argument when launching vim (e.g. `$ vim test.go`) and launching vim with no file to edit then opening the file with `:e`. I do see that `vim-go` gets successfully added to vim's `runtimepath` but only the commands listed in the orignal message gets loaded.

```
runtimepath=~/.vim,~/.vim/pack/dist/opt/vim-go,~/.vim/pack/dist/opt/deoplete-go,~/.vim/pack/dist/opt/gotests-vim
```

**Comment by bhcleek** [comment]
> I tried executing it 3 different ways, the end goal was if I'm not working on a Go project then vim should skip this plugin.

This generally sounds like a premature optimization. Vim-go is is technically two plugins: a global plugin which provides the commands you're seeing available (e.g. `:GoUpdateBinaries, `:GoPath, `:GoInstallBinaries`, etc.) and a file type plugin that is only loaded when the file type is set to `go`.  When you're not in a Go file, vim-go is a global plugin and has _very_ low cost. I recommend not worrying about trying to keep vim-go from loading when you're not dealing with Go files.

> The first way was having packadd vim-go in $HOME/.vim/ftplugins/go.vim, when that didn't work I put it is $HOME/.vim/after/ftplugns/go.vim (like how it currently is in https://gitlab.com/vendion/dotfiles/blob/opt-vim-go/.vim/after/ftplugin/go.vim), lastly when neither of those worked I tried calling it manually from within vim.
> I tried calling them after opening a Go file, by passing the file name as an argument when launching vim (e.g. $ vim test.go) and launching vim with no file to edit then opening the file with :e. I do see that vim-go gets successfully added to vim's runtimepath but only the commands listed in the orignal message gets loaded.

`ftplugin/go.vim` is only the first bit of vim-go. There are many other files needed, some of which have been copied into the files provided by vim itself; those files are generally old and won't provide all the features you'd expect from your installed vim-go, so by putting vim-go's files in an `after` directory, they're likely to be loaded too late because of how Vim orders the loading of plugins and filetype plugins from the runtimepath entries.

Vim-go generally works fine in a `pack/opt` directory, but you have to load it at the right time to get the functionality you expect, because many commands are not setup until the filetype is set; if you try loading vim-go after the filetype is set, you're not going to get all of vim-go's functionality.

**Comment by vendion** [comment]
> This generally sounds like a premature optimization. Vim-go is is technically two plugins: a global plugin which provides the commands you're seeing available (e.g. :GoUpdateBinaries, :GoPath, :GoInstallBinaries, etc.) and a file type plugin that is only loaded when the file type is set to go. When you're not in a Go file, vim-go is a global plugin and has very low cost. I recommend not worrying about trying to keep vim-go from loading when you're not dealing with Go files.

This may full under premature optimization, but for me it's more of keeping things clean and minimal. While it doesn't affect vim's performance but it bothers me to have plugins cluttering my `runtimepath` with plugins related to projects I'm not working on (i.e. I don't really need PHP or Perl related plugins in my `runtimepath` if I'm working on a Go project and vice versa) sure they are not doing anything since the filetype wont match them. Nor do I really want all my plugins in one massive directory which is all problems I have with other plugin managers (pathogen, vim-plug, etc...). This is something I'm hoping I'll be able to solve with Vim8 packages by having multilple directories under _$HOME/.vim/pack_  such as:

```
$HOME/.vim/pack/dist/{opt,start} <- General plugins and theme that is always used
$HOME/.vim/pack/go/{opt,start} <- Plugins just related to Go development
$HOME/.vim/pack/php/{opt,start} <- Plugins just related to PHP development
...
```

> ftplugin/go.vim is only the first bit of vim-go. There are many other files needed, some of which have been copied into the files provided by vim itself; those files are generally old and won't provide all the features you'd expect from your installed vim-go, so by putting vim-go's files in an after directory, they're likely to be loaded too late because of how Vim orders the loading of plugins and filetype plugins from the runtimepath entries.

Since _$HOME/.vim/ftplugin_ and _$HOME/.vim/after/ftplugin_ are too late in the process, would creating a `ftdetect` with a autocommand for the `.go` file extension be soon enough?

> Vim-go generally works fine in a pack/opt directory, but you have to load it at the right time to get the functionality you expect, because many commands are not setup until the filetype is set; if you try loading vim-go after the filetype is set, you're not going to get all of vim-go's functionality.

I would argue that since the plugin needs to be loaded before the filetype is set in Vim, Vim-go may work but I wouldn't say it does so "fine".

**Comment by bhcleek** [comment]
TL;DR: if you want to put vim-go into an optional package, then you'll need to make sure to load it with `packadd` before opening a Go file.

> Since $HOME/.vim/ftplugin and $HOME/.vim/after/ftplugin are too late in the process, would creating a ftdetect with a autocommand for the .go file extension be soon enough?

I don't see a way to leverage an ftdetect file to bend vim-go to your desired configuration unless I'm misunderstanding what you're suggesting. ftdetect should only contain an autocmd that sets the filetype. Both `:help ftdetect`, and https://learnvimscriptthehardway.stevelosh.com/chapters/42.html#vimftdetect are relevant. You can see in [vim-go's source](https://github.com/fatih/vim-go/blob/master/ftdetect/gofiletype.vim#L8) and [vim's source](https://github.com/vim/vim/blob/master/runtime/filetype.vim#L666) that the filetype is set as expected (https://github.com/vim/vim/blob/master/runtime/filetype.vim#L2184 is also topical).

> I would argue that since the plugin needs to be loaded before the filetype is set in Vim, Vim-go may work but I wouldn't say it does so "fine".

The bulk of vim-go _is_ a filetype plugin and therefore acts when the filetype is set. It works as intended for the majority of users and isn't going to be refactored to satisfy outliers without a a solid reason.

**Comment by bhcleek** [comment]
@vendion should we keep this open?

**Comment by vendion** [comment]
It's looking like this can be closed as it seems there is no point in having `vim-go` to be in the pack/*/opt directory due to how it would have to be loaded so early, It might as well just be the pack/*/start directory.
